### PR TITLE
Kitchen2.0

### PR DESCRIPTION
**Pull request by mir7b**

### CODE REVIEW

**Comment by gaya-** [comment]
Awesome!. I have one request before testing this. Could you please merge your kitchen2 directory with iai_apartment? You could simply include your kitchen.urdf in the apartment.urdf, that would be nice and modular. We would spawn the robot always in the complete apartment with walls, couches and the kitchen, instead of only using the kitchen.

**Comment by mir7b** [comment]
Done.

File: iai_apartment/launch/kitchen2_display.launch [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<launch>
+
```
Comment (ichumuh) [inline iai_apartment/launch/kitchen2_display.launch:2]: There should be a static transform publisher, which links the apartment to map

    <node pkg="tf"
        type="static_transform_publisher"
        name="iai_apartment_room_link_broadcaster" 
        args="0.0 0.0 0. 0.0 0. 0. map iai_apartment/kitchen_root 25">
    </node>

File: iai_apartment/launch/kitchen2_display.launch [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+	<node name="apartment_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher">
+		<remap from="robot_description" to="apartment_description" />
+		<remap from="joint_states" to="apartment_joint_states" />
+		<param name="tf_prefix" value="iai_apartment" />
+		<param name="publish_frequency" value="60" />
+	</node>
+	<node name="joint_state_publisher_gui" pkg="joint_state_publisher_gui" type="joint_state_publisher_gui">
+		<remap from="robot_description" to="apartment_description" />
+		<remap from="joint_states" to="apartment_joint_states" />
+	</node>
```
Comment (ichumuh) [inline iai_apartment/launch/kitchen2_display.launch:14]: Only use this in the _display.launch files. In the _upload.launch, this should be replaced with

    <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher">
        <remap from="robot_description" to="apartment_description" />
        <remap from="joint_states" to="apartment_joint_states" />
        <rosparam param="source_list">
            - /kitchen/cram_joint_states
        </rosparam>
    </node>

the source list allows us to set the joint state of the kitchen via topic

**Review by ichumuh — COMMENTED** [review]
- I'd split the launch files, one that is called upload, which starts static transform publisher, joint_state_publisher, robot_state_publisher and uploads the urdf. Then another one which uses joint_state_publisher_gui and rviz. Usually we'll only need the first launch files and display is just to get a quick look
- the apartment.urdf should include the kitchen urdf, as @gaya- suggested

**Comment by ichumuh** [comment]
- This is debatable, but I think the joints should all open in the positive direction

**Comment by ichumuh** [comment]
I can not reach into the collision meshes with giskard. Bullet needs convex decompositions for the collision meshes: https://github.com/bulletphysics/bullet3/issues/1507
Can you please change that?

**Comment by ichumuh** [comment]
an alternative would be to give the collision entries multiple box geometries, similar to how it is done for the tiago: https://github.com/pal-robotics/tiago_robot/blob/ee8c909be59aec65cf707c02013cfed37be2d051/tiago_description/urdf/torso/torso.urdf.xacro#L116

**Comment by gaya-** [comment]
> * This is debatable, but I think the joints should all open when the joint state increases

Yes, I require this as well.

**Comment by gaya-** [comment]
Could you also make the handle coordinate frames as defined here:
https://ai.uni-bremen.de/wiki/3dmodeling/items

**Comment by gaya-** [comment]
Meh, convex hull of the apartment walls is the whole apartment because it's one mesh x_x. So everything is in a collision. Looking forward to @ichumuh's obj meshes.

Otherwise everything else looks good to me.

Final todos:

- Jeroen wanted to provide a more-or-less ok position for the kitchen in the apartment, but it will have to be tweaked when it's installed anyway. So I'd say at the left (when entering through the door left) wall centered with the wall would be ok for now.
- There is still kitchen2_distplay.launch that has Simon's comments on it. Either remove the launch file or address the comments.
- I noticed links like "sa" "sb" "side1" etc., for sockets I guess. Would be good if they had more explicit names.
- This is a bit unrelated, but it would be cool if you centered the apartment such that the global map origin is in the middle of the room. @Sanic should we change the origin in the Unreal mesh or rather in the URDF? The mesh has a weird origin, doesn't it?

**Comment by gaya-** [comment]
Ah, yes. The difficult part. To describe a location inside the top left cupboard I need a link for the cupboard, or at least the bottom board of cabinet1 top part needs to be its own link. Otherwise I cannot explain to the robot where the cup is using the URDF links that already exist. Is it possible to cut off parts from a mesh and separate it into two meshes? With FreeCAD or something?
I guess I'll request this in a separate PR.

**Comment by mir7b** [comment]
Should be okay to check out now.

**Comment by mir7b** [comment]
> I noticed links like "sa" "sb" "side1" etc., for sockets I guess. Would be good if they had more explicit names.

considering those have no functionality, I will leave the proper naming of them to the final tweaks. Side A and side B are the two sides of the kitchen, I insist on that name staying because they make life easier in tweaking. 

> This is a bit unrelated, but it would be cool if you centered the apartment such that the global map origin is in the middle of the room.

Check out after this commit. this should be fixed now. 

> There is still kitchen2_distplay.launch that has Simon's comments on it. Either remove the launch file or address the comments.

Check the new commit.

>  more-or-less ok position for the kitchen in the apartment, 

it's open for tweak. one just needs to know their way around a URDF to tweak it. 

> Could you also make the handle coordinate frames as defined here:

done.

> To describe a location inside the top left cupboard 

I could have done that easily at the beginning by changing their centre in blender, and then putting them in the kitchen URDF. But now this will cause a lot of rework. Let's discuss this later and find a solution. 

> I think the joints should all open when the joint state increases

fixed now. you might want to change the wardrobe that's already in there though. 

> display.launch file 

already removed and merged as xacro in the apartment. 

is there anything else to be addressed?

**Comment by mir7b** [comment]
Ooops! Accidentally hit "close" instead of "comment"!

**Comment by Sanic** [comment]
> This is a bit unrelated, but it would be cool if you centered the apartment such that the global map origin is in the middle of the room. @Sanic should we change the origin in the Unreal mesh or rather in the URDF? The mesh has a weird origin, doesn't it?

Having this consistent would be great!
I can send @mir7b the Link to the UE4 apartment. I think that the origin in the UE4 model should be close to the center of the room.

**Comment by gaya-** [comment]
Can we make it in a corner? Like the left left corner near the kitchen? It is like that in the refills lab, makes physically measuring stuff in the real world easier.

**Comment by gaya-** [comment]
The origin is below ground.

**Comment by Sanic** [comment]
The origin should be directly on the floor level, or?

**Comment by ichumuh** [comment]
what about the tf static transform publisher and the joint state publisher in the bringup.launch?

**Comment by gaya-** [comment]
The bringup.launch from Giang?

**Comment by gaya-** [comment]
Can you add what you need in the obj PR? I have my own launch file, so I don't care.

**Comment by mir7b** [comment]
> what about the tf static transform publisher and the joint state publisher in the bringup.launch?

doesn't need it. you can use it the same way you use the apartment. It is merged in there anyways.

**Pull request merged**